thetis/callback.py

```diff
@@ -554,7 +569,13 @@ def message_str(self, *args):
             for name, values in zip(self.detector_names, args))
 
     def _evaluate_field(self, field_name):
-        return self.solver_obj.fields[field_name](self.detector_locations)
+        field = self.solver_obj.fields[field_name]
+        P0DG, P0DG_input_ordering = self.function_spaces[field_name]
+        f_at_points = Function(P0DG)
+        f_at_input_points = Function(P0DG_input_ordering)
```

Comment:
These function are created on every evaluation. Instead of storing the function spaces in `self.function_spaces` you could just store the functions themselves.

Reply:
This is much more sensible